src/resource_manager/resource_manager.py

```diff
@@ -0,0 +1,250 @@
+
+        Args:
+            resource: The resource object to create.
+        """
+
+    @abc.abstractmethod
+    def _patch_resource(self, name: str, resource: AnyResource) -> None:
+        """Abstract method to patch an existing resource in a given namespace.
+
+        Args:
+            name: The name of the resource to patch.
+            resource: The modified resource object.
+        """
+
+    @abc.abstractmethod
+    def _list_resource(self) -> typing.List[AnyResource]:
+        """Abstract method to list resources in a given namespace based on a label selector."""
+
+    @abc.abstractmethod
+    def _delete_resource(self, name: str) -> None:
+        """Abstract method to delete a resource from a given namespace.
+
+        Args:
+            name: The name of the resource to delete.
+        """
+
+    def define_resource(
+        self,
+        definition: ResourceDefinition,
+    ) -> AnyResource:
+        """Create or update a resource in kubernetes.
+
+        Args:
+            definition: The ingress definition
+
+        Returns:
+            The name of the created or modified resource.
+
+        Raises:
+            InvalidResourceError: If the generated resource is invalid.
+        """
+        resource_list = self._list_resource()
+        resource = self._gen_resource_from_definition(definition)
+        res_name = resource_name(resource)
+        if not res_name:
+            raise InvalidResourceError("Missing resource name.")
+
+        resources = [resource_name(r) for r in resource_list if resource_name(r) is not None]
+        if res_name in resources:
+            self._patch_resource(name=res_name, resource=resource)
+            logger.info(
+                "%s updated in namespace %s with name %s",
+                self._name,
+                self._namespace,
+                res_name,
+            )
+        else:
+            self._create_resource(resource=resource)
+            logger.info(
+                "%s created in namespace %s with name %s",
+                self._name,
+                self._namespace,
+                res_name,
+            )
+        return resource
+
+    def cleanup_resources(
+        self,
+        exclude: typing.Optional[AnyResource] = None,
+    ) -> None:
+        """Remove unused resources.
+
+        Args:
+            exclude: The name of resource to be excluded from the cleanup.
+        """
+        for resource in self._list_resource():
+            excluded_resource_name = resource_name(exclude)
+            res_name = resource_name(resource)
+            if not res_name or not excluded_resource_name:
+                continue
+            if res_name == excluded_resource_name:
+                continue
+            self._delete_resource(name=res_name)
+            logger.info(
+                "%s deleted in namespace %s with name %s",
+                self._name,
+                self._namespace,
+                resource,
+            )
```

Comment:
By filtering by name 2 deployments will silently compete, Weii suggested to add the relation ID or the app name in the resource name to distinguish

Reply:
An alternative would be to check (if available) if there are any other deployment of the same charm